syntax = "proto3";

package library;

import "google/protobuf/any.proto";

service Library {
    // Add book to mongodb (only librarians can do that)
    rpc add_book (Book) returns (Message);
    
    // Remove a book by name from the DB (only librarians can do that)
    rpc delete_book (BookName) returns (Message);

    // updates some fields of a book (only librarians can do that)
    rpc update_book (UpdatedBook) returns (Message);

    // gets a book by name
    rpc get_book (BookName) returns (Book);

    // returns books sorted in alphabetical order implementing pagination
    rpc get_books_list (Pagination) returns (Books);
}


message Book {
    string name = 1;
    repeated string authors = 2;
    string publisher = 3;
    string date_of_release = 4;
    enum Genre {
        NONE = 0;
        SCIENCE_FICTION = 1;
        ROMANCE = 2;
        DRAMA = 3;
        COMEDY = 4;
        HORROR = 5;
        ADVENTURE = 6;
        ACTION = 7;
    }
    Genre genre = 5;
    string description = 6;
}


message Books {
    repeated Book book = 1;
}

message Message {
    string message = 1;
}


message BookName {
    string name = 1;
}

message Pair {
    enum Fields {
        NONE = 0;
        name = 1;
        authors = 2;
        publisher = 3;
        date_of_release = 4;
        genre = 5;
        description = 6;
    }

    Fields key = 1;
    google.protobuf.Any value = 2;
}

message UpdatedBook {
    repeated Pair updated_fields = 1;
}

message Pagination {
    int32 offset = 1;
    int32 limit = 2;
}
